Neos.Flow/Classes/Http/Middleware/SecurityEntryPointMiddleware.php

```diff
@@ -0,0 +1,86 @@
+<?php
+declare(strict_types=1);
+
+namespace Neos\Flow\Http\Middleware;
+
+use GuzzleHttp\Psr7\Response;
+use Neos\Flow\Annotations as Flow;
+use Neos\Flow\Log\Utility\LogEnvironment;
+use Neos\Flow\Security\Authentication\Token\SessionlessTokenInterface;
+use Neos\Flow\Security\Authentication\TokenInterface;
+use Neos\Flow\Security\Context;
+use Neos\Flow\Security\Exception\AuthenticationRequiredException;
+use Psr\Log\LoggerInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+/**
+ * A HTTP middleware that handles authentication exceptions that were thrown by the dispatcher (@see \Neos\Flow\Mvc\Dispatcher::dispatch()) and
+ * * rethrows the exception if no token with Entry Point is authenticated
+ * * or otherwise invokes the Entry Point of all authenticated tokens
+ */
+class SecurityEntryPointMiddleware implements MiddlewareInterface
+{
+    /**
+     * @Flow\Inject
+     * @var Context
+     */
+    protected $securityContext;
+
+    /**
+     * @Flow\Inject(name="Neos.Flow:SecurityLogger")
+     * @var LoggerInterface
+     */
+    protected $securityLogger;
+
+    /**
+     * @inheritDoc
+     */
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $next): ResponseInterface
+    {
+        try {
+            return $next->handle($request);
+        } catch (AuthenticationRequiredException $authenticationException) {
+            /** @var TokenInterface[] $tokensWithEntryPoint */
+            $tokensWithEntryPoint = array_filter($this->securityContext->getAuthenticationTokens(), static function (TokenInterface $token) {
+                return $token->getAuthenticationEntryPoint() !== null;
+            });
+
+            if ($tokensWithEntryPoint === []) {
+                $this->securityLogger->notice('No authentication entry point found for active tokens, therefore cannot authenticate or redirect to authentication automatically.');
+                throw $authenticationException;
+            }
+
+            $response = $this->buildHttpResponse();
```

Comment:
Creating a new response (through a method for test-friendlyness), because we caught an exception, so there is no valid response built in the inner layer.